src/examples/signal_generator.cpp

```diff
@@ -0,0 +1,140 @@
+                            "}";
+
+  //input_proto = "grpc";
+
+  // Dynamic Property
+  signal_generator_sm.init(ComponentConfiguration(generator_configuration));
+  digital_converter_sm.init(ComponentConfiguration(converter_configuration));
+
+  auto remote_endpoint = ConnectionConfiguration("{ analog_input: { endpoint: '/generator/analog_output' } }");
+
+  signal_generator_sm.start();
+  digital_converter_sm.start();
+
+  signal_generator_sm.connect();
+  digital_converter_sm.connect(remote_endpoint);
+
+  signal_generator_sm.check();
+  digital_converter_sm.check();
+
+  bool stop = false;
+  //std::atomic<bool> stop(false);
+
+  /* This thread is done for getting the user's input while it can be plotted */
+  std::thread input_thread([&signal_generator_sm, &digital_converter_sm, &stop]() {
+    float amp, freq;
+    bool cosine;
+    int option;
+    int cosine_option;
+    /* Dynamic properties Access */
+    DynamicPropertyAccess dynamic_property_access("/generator"_uri);
+    while (!stop)
+    {
+      /* Read parameters values */
+      dynamic_property_access.get<float>("amplitude", amp);
+      dynamic_property_access.get<float>("frequency", freq);
+      dynamic_property_access.get<bool>("cosine", cosine);
+
+      std::cout << "Current parameters" << std::endl;
+      std::cout << "Amplitude: " << amp
+                << " Frequency: " << freq
+                << " Waveform: " << (cosine ? "Cosine" : "Sine")
+                << std::endl;
+      std::cout << "------------------------------------------------" << std::endl
+                << SET_VALUES_OPTION << " - Choose new values" << std::endl
+                << EXIT_OPTION << " - Quit" << std::endl
+                << "-> ";
+
+      std::cin >> option;
+
+      if (option == SET_VALUES_OPTION)
+      {
+        std::cout << "Set Amplitude: ";
+        std::cin >> amp;
+        std::cout << "Set Frequency: ";
+        std::cin >> freq;
+        std::cout << "1 - Cosine " << std::endl
+                  << "2 - Sine " << std::endl
+                  << "-> ";
+        std::cin >> cosine_option;
+        cosine = ((cosine_option == 1) ? true : false);
+        dynamic_property_access.set<float>("amplitude", amp);
+        dynamic_property_access.set<float>("frequency", freq);
+        dynamic_property_access.set<bool>("cosine", cosine);
+      }
+      stop = (option == EXIT_OPTION)?true:false;
+    };
+  });
+
+  std::thread th1([&signal_generator_sm, &stop]() {
+    while (!stop)
+    {
+      signal_generator_sm.update();
+      std::this_thread::sleep_for(std::chrono::milliseconds(100));
+    }
+  });
+
+  widgets::registerChannelListener<hyro::SignalMsgs>("protocol: 'grpc', ip: '0.0.0.0',   port: '50051'", "grpc", [](hyro::SignalMsgs msg) {
+    widgets::plot2d<float>("Analog Signal", "/generator/analog_output", msg.value);
+  });
+
+  widgets::plot2d<float>("Digital Signal", "/converter/digital_output", widgets::Plot2dSettings::initWithProtocol("api"));
+
+  if(!stop)
```

Comment:
I think this code does not work.

The code `widgets::exec` make the program stopped here until you close all the widgets. The easiest way to make your algorithm works is moving `widgets::exec` in the thread and move the code inside `input_thread` to the main thread.

Another option is remove this verification.

Reply:
I tested it. It is working.